### ISSUE TITLE
Experiment with a "copy small blocks" strategy

### ISSUE THREAD

**Issue by lemire** [lemire/simdjson#442]
We divide documents into small (e.g., 16kB) chunks. The document is copied (in units of say 16kB) where it may be padded as needed. We run stage 1 over it, then stage 2. Then we continue parsing with the next block.

This would solve https://github.com/lemire/simdjson/issues/174 and might improve performance over large documents.

**Comment by aqrit** [comment]
memory-mapped ring buffer, perhaps.

**Comment by lemire** [comment]
Note that this would reduce the number of pages that need to be allocated, which is a bottleneck in large files: https://github.com/lemire/simdjson/pull/443

**Comment by jkeiser** [comment]
I'm all for processing the document in chunks, even in stage 1. It would make it more feasible to cut things into *more* stages that do less work, getting rid of even more branches (all strings in one place, all numbers in another, etc.).

One thing I don't understand: why copy instead of passing the 16KB buffers directly (at least until the last one, which you'd have to copy and fill)? Won't copying cost 1 cycle every 32 bytes on AVX2? Given we're at 90-100 cycles per 64 bytes on AVX2, that's a 2% throughput increase, so not the worst in the world, especially if we can get it back elsewhere as a result.

One alternative is to emit "end value" indices for strings, numbers and atoms, so that stage 2 doesn't have to scan past the end of a value. This alternative is only reasonable if we switch to a [SIMD index wrrite](https://lemire.me/blog/2018/03/08/iterating-over-set-bits-quickly-simd-edition/), however, because we'd be increasing the index density by up to 2x. It's going to increase memory writes as well, but definitely shaves instructions off the read end. You also have to figure out what to do with the very last index, and might have to do some special handling for `\u`.

**Comment by lemire** [comment]
@jkeiser 
I think that I explain the rationale for a copy: it is for the padding and whatever surgery you need to do at the beginning and end of the segment. Of course, it would be best to avoid a copy if we can. I will change the title to reflect that.

**Comment by lemire** [comment]
> One alternative is to emit "end value" indices for strings, numbers and atoms, so that stage 2 doesn't have to scan past the end of a value. 

If we could do that without massively harming the performance of stage 1, it might be nice.

What we could never figure out is how much of a gain it could be in stage 2. That is, would stage 2 be much easier if we had the start and end of the atoms?

**Comment by jkeiser** [comment]
I don't think it would be much easier if it had the start and end of atoms in particular ... it would save a few instructions where we check for whitespace or structural, but those are largely parallel instructions in a branchy place that is likely to be able to run them at max parallelism (one of the effects of a branch misprediction is it clears out all execution core contention). It might be a wash, unless removal of the padding code became possible and helped a lot.

I'll lay out what I'd really like extra indices for in a separate issue to stop hijacking this one :)

**Comment by lemire** [comment]
Delaying again to 0.7.

**Comment by jkeiser** [comment]
Same question as #670 : do we need input streaming for 1.0?

**Comment by lemire** [comment]
Maybe not.

**Comment by jkeiser** [comment]
Moving out to 1.0 until we decide whether or not to block on it.

**Comment by dongx-psu** [comment]
I think it is a good feature to have padded on request only on the final small block. This will definitely help to get better parsing performance not only on large documents but also on document batches.
